### ISSUE TITLE
"Requires" and concepts

### ISSUE THREAD

**Issue by DarkerStar** [foonathan/standardese#8]
One of the examples is:

~~~cpp
/// \effects Exchanges values stored in two locations.
/// \requires Type `T` shall be `MoveConstructible` and `MoveAssignable`.
template <class T>
void swap(T &a, T &b) noexcept(...);
~~~

Perhaps it would be better to have the "requires" clause formatted in some way other than prose. For example:

~~~cpp
/// \effects Exchanges values stored in two locations.
/// \requires T MoveConstructible
/// \requires T MoveAssignable
template <class T>
void swap(T &a, T &b) noexcept(...);
~~~

Or:

~~~cpp
/// \effects Exchanges values stored in two locations.
/// \requires T MoveConstructible MoveAssignable
template <class T>
void swap(T &a, T &b) noexcept(...);
~~~

Both of these could produce exactly the same result as the original.

However, it would be easy to add enhancements, like linking to concept documentation. In the long run, these manual "requires" can even be combined with stuff gleaned from in-code `requires` expressions when concepts rolls around, and maybe even eventually phased out:

~~~cpp
// TODAY:
/// \effects Exchanges values stored in two locations.
/// \requires T MoveConstructible MoveAssignable
template <class T>
void swap(T &a, T &b) noexcept(...);

// C++20 (manual requires command redundant, but not an error):
/// \effects Exchanges values stored in two locations.
/// \requires T MoveConstructible MoveAssignable
template <class T>
requires MoveConstructible<T> && MoveAssignable<T>
void swap(T &a, T &b) noexcept(...);

// C++?? (unnecessary command eventually removed):
/// \effects Exchanges values stored in two locations.
template <class T>
requires MoveConstructible<T> && MoveAssignable<T>
void swap(T &a, T &b) noexcept(...);
~~~

Manually added requires clauses can be added with requires clauses deduced from code (when concepts rolls around):

~~~cpp
/// \requires T MoveAssignable
/// \requires "Something else."
template <typename T>
requires DefaultConstructible<T>
void foo(T);

// produces:
Requires: `T` must be MoveAssignable and DefaultConstructible. Something Else.

/// \requires "And so on."
void foo(EqualityComparable a);

// produces:
Requires: Type of argument `a` must be EqualityComparable. And so on.
~~~

We don't have concepts yet, but we know they're coming. Having a way to express concept requirements in documentation *today*... that we can later convert to *actual* concepts checked by the compiler sometime in the future... would be awesome.

If the command "`\requires`" cannot be changed (for compatibility with Doxygen), then perhaps a new command "`\requires-concept`" or "`\requires-constraints`" could be used.

**Comment by foonathan** [comment]
Compatibility with Doxygen is no concern.

I haven't thought about a shorthand syntax, is a great idea.
But not for \requires because it will also be used for preconditions so parsing will get difficult.
Maybe a \requires-concepts like you mentioned.

I do plan on automatically generating the \requires clause, maybe opt-in through `\requires auto` or similar.
This will definitely be implemented with concept support but I also try to detect SFINAE parameters like so:

```cpp
template <typename T, typename = std::enable_if_t<std::is_move_constructible<T>::value>>
void func(T t);
```

To generate a:

Requires: `std::is_move_constructible<T>::value` must be `true`.

Or even expression SFINAE:

```cpp
template <typename T, typename = decltype(std::declval<T>().foo())>
void func(T t);
```

Requires: The expression `std::declval<T>().foo()` must be well-formed.

Maybe I'll pull of something like:

Requires: For an object `t` of type `T` the expression `t.foo()` must be well-formed.

**Comment by Manu343726** [comment]
I know clang 3.7 has no support for concepts, this is just an idea for future work: Like you suggested getting concepts from SFINAE (+1 from me), let me celebrate the release of GCC 6.1 here:

``` cpp
void func(MoveConstructible mc);
```

Requires: Parameter mc must model MoveConstructible concept.

**Comment by timsong-cpp** [comment]
*Requires:* clauses in the standard and the concepts `requires` constraints are different beasts. Breaking a *Requires:* clause results in undefined behavior unless otherwise specified; not satisfying a constraint means that the overload doesn't participate in overload resolution, but the behavior is well-defined. 

The equivalent of constraints and `enable_if<foo>` in standardese is "*Remarks*: This function shall not participate in overload resolution unless foo".

**Comment by foonathan** [comment]
Thanks for the info.

**Comment by DarkerStar** [comment]
I don't think the standard is particularly consistent about how it denotes what would be used as concepts constraints. For example, swap (20.2.2) in the [C++17 draft standard](http://open-std.org/JTC1/SC22/WG21/docs/papers/2016/n4594.pdf) has:

> *Requires:* Type `T` shall be `MoveConstructible` (Table 20) and `MoveAssignable` (Table 22).

Which pretty much exactly what you'd write with concepts.

Adding to the complexity is the [proposed *Preconditions:* clause](http://open-std.org/JTC1/SC22/WG21/docs/papers/2016/p0411r0.html), which is supposed to take preconditions away from *Requires:*. (And things will probably get even *more* complex if [contracts](http://open-std.org/JTC1/SC22/WG21/docs/papers/2016/p0380r1.pdf) get implemented.)

But @timsong-cpp has an interesting point! Perhaps the best way to get closest to what the standard does is to detect if SFINAE is being used to prevent the function/type from being recognized as a valid match in overload resolution, and if so put it in the *Remarks:* section (as "shall not participate"), otherwise put it in *Requires:* (or *Preconditions:*).

For example:

```cpp
template <typename T, typename = std::enable_if_t<std::is_move_constructible_v<T>>>
void func1(T t);

template <typename T>
requires MoveConstructible<T>
void func2(T t);

template <MoveConstructible T>
void func3(T t);

void func4(MoveConstructible t);
```

`func1` would generate a *Remarks:* clause that says "shall not participate" unless `std::is_move_constructible_v<T>` is `true`.

`func2`, `func3`, and `func4` would all generate a *Requires:* clause that says `T` shall be `MoveConstructible`.

(Although, technically, Concepts does the same thing as SFINAE - the concepts remove the functions from overload resolution.)

*Or* you could use SFINAE and Concepts - anything that removes a function/type from overload resolution - to generate the *Remarks:* (with "shall not participate"), and use `static_assert`s (and possibly, later, something like the proposed `[[expects]]`) to generate the *Requires:*. That's probably closest to the spirit of the standard's clauses, if not necessarily how they're actually used.

Or you could just say, meh, given the choice between matching the standard exactly and generating documentation easily readable by people who aren't standards wonks, the latter is probably more useful. *Requires:* just makes more sense for requirements on types than *Remarks:*, and in practice the difference between SFINAE removing a function from overload resolution for a given set of arguments (assuming no other matches exist) versus a hard error isn't really that important; in either case the message to the user is "it won't work". And if there *are* other matches, then the message for the match(es) that do(es) work is "but this will". Which is more or less how we already document overloads.

(Of course the *best* of all possible solutions would be if the standard were to reorganize its clauses to better match the language - a clause for type requirements (ideally, *Requires:*, to match the concepts language), a clause for preconditions (*Preconditions:*, you could probably generate something similar from something like `[[expects]]`, or `std::assert`s, `static_assert`s and `[[assert]]`s at that start of a function), a clause for postconditions (*Postconditions:*, you could use `[[ensures]]` and `std::assert`s just before returns). Perhaps once the standard has concepts and contracts, that might be worth proposing. Wouldn't it be neat if standardese actually influenced a change in the standard itself? ^_^;)

**Comment by timsong-cpp** [comment]
> I don't think the standard is particularly consistent about how it denotes what would be used as concepts constraints. For example, swap (20.2.2) in the C++17 draft standard has:
> 
>     *Requires*: Type `T` shall be `MoveConstructible` (Table 20) and `MoveAssignable` (Table 22).
> 
> Which pretty much exactly what you'd write with concepts.
> 

There's no actual inconsistency. The named requirements `MoveConstructible` and `MoveAssignable` include a semantic requirement that move construction and move assignment "behave as advertised", i.e., that `A a(std::move(b));` and `a = std::move(b)` actually makes `a`'s state equivalent to `b`'s previous state.

[The preceding *Remarks* clause](https://timsong-cpp.github.io/cppwp/utility.swap#1) is what actually requires SFINAE on swap, so that the new `is_swappable` traits can work.
